### PR TITLE
Issue 1406: Allow unused variable in for loop iterator

### LINKED ISSUE #1406: WPS122 should allow a no-op loop var
# Bug report

## What's wrong

Imagine I want to do something 3 times but I don't care about the attempt number. So I'd normally write something like this:
```python
for _ in range(3):
    do_something()
```

But then the linter complains with `WPS122 Found all unused variables definition: _` which is incorrect. While things like assignment allow to fix this with
```diff
-_something_unused = get_useless_result()
+get_useless_result()
```
it is impossible with for-loops not to define this loop variable at all.

## How is that should be

This case shouldn't be a violation of the given rule.

## System information

N/A

### PR DESCRIPTION
**Pull request by nn3un**
# I have made things!

## Checklist

- [x] I have double checked that there are no unrelated changes in this pull request (old patches, accidental config files, etc)
- [ ] I have created at least one test case for the changes I have made
   (Moved a test case from unallowed to allowed)
- [ ] I have updated the documentation for the changes I have made
   (The [documentation](https://wemake-python-stylegui.de/en/latest/pages/usage/violations/naming.html#wemake_python_styleguide.violations.naming.UnusedVariableIsDefinedViolation) makes no mention of this specific case)
- [ ] I have added my changes to the `CHANGELOG.md`

## Related issues
Closes #1406

(The issue description talks about comprehensions, but I didn't find any place,including the [comprehension tests](https://github.com/wemake-services/wemake-python-styleguide/tree/58b96219a39390c35353aa9a29d483d5ed3f1cf7/tests/test_visitors/test_ast/test_loops/test_comprehensions), where comprehensions were being checked for WPS122, and manually testing code such as `my_list = [3 for _ in range(4)]` was throwing no WPS122 errors, so I ended up doing nothing for that)

### CODE REVIEW

File: tests/test_visitors/test_ast/test_naming/conftest.py [hunk 1]
```diff
@@ -211,7 +211,6 @@ def container():
     unpacking_variables,
     variable_def,
     with_variable,
-    for_variable,
```
Comment (sobolevn) [inline tests/test_visitors/test_ast/test_naming/conftest.py:214]: Can you please add a test that explicitly shows that this is now allowed?

**Comment by nn3un** [comment]
Added tests that test the complement of _FORBIDDEN_RAW_UNUSED_TEMPLATE, which includes for loops. So now tests run for: 

```
for _ in []:
for __ in []:
for _unused in []:
```
in method, function, and module contexts.

(Please let me know if the docstrings need to be improved.)

**Comment by sobolevn** [comment]
@nn3un here's when I start to see a problem:
1. `for _ in []:` is allowed
2. `for __ in []:` is not allowed
3. `for _unused in []:` is not allowed

**Comment by nn3un** [comment]
> @nn3un here's when I start to see a problem:
> 
> 1. `for _ in []:` is allowed
> 2. `for __ in []:` is not allowed
> 3. `for _unused in []:` is not allowed

Oh, I didn't realize that. 
- `for __ in range(n)` still raises  ` WPS123 Found wrong unused variable name: __`. So is it ok if it doesn't raise WPS122? I think it will make the code a bit cleaner, since I can call [access.is_unused](https://github.com/wemake-services/wemake-python-styleguide/blob/256de41bebbe53f60cd5ce87cb4031f120eb9615/wemake_python_styleguide/logic/naming/access.py#L9-L26), which returns `True` for both _ and __.

- Should `for _, _ in result:` raise error?

**Comment by sobolevn** [comment]
Yes to both questions 👍

**Comment by nn3un** [comment]
Sorry for the delay. I tried different ways to change the tests to accommodate for the change, and the best I could think of required a decent bit of restructuring. 

My basic idea was unused variables are of two kinds: raw (`_`,`__`) and protected (`_protected`). So as long as I checked for raw forbidden, raw allowed, protected forbidden, and protected allowed, I checked for all unused definition use cases.

Git makes the `test_unused_defintion.py` changes a bit complex to understand so here’s a summary of what I did:

The 2 tuple tests, `test_used_variable_definition` and `test_raw_unused_variable_definition` remain unchanged (with slight comment changes and renaming)

Deleted `test_unused_variable_definition_allowed`, which is now covered by `test_protected_unused_variable_definition_allowed`, and `test_raw_unused_variable_definition_allowed`. 

`test_unused_variable_definition_allowed` is now `test_protected_unused_var_definition_allowed`. 

Added `test_raw_unused_variable_definition_allowed` (which tests for the no-op loop var)

**Comment by nn3un** [comment]
Hurrah! Thank you @sobolevn for all the help and guidance!

**Pull request merged**